### PR TITLE
Add PathSpecSet: a set of PathSpecSet with support for full projection

### PR DESCRIPTION
**Pull request by GuiForget**
The PathSpecSet provides functionalities to more easily build or edit a set of PathSpec. In addition, it offers the support for a full projection so that methods like "contains" always return true making it easier for the client code to write logic with the set.

### CODE REVIEW

**Comment by junchuanwang** [comment]
Hey @GuiForget Is this still a draft or published?

**Comment by GuiForget** [comment]
@junchuanwang it's still a draft, I need to finish the unit tests. I made some progress a couple of weeks ago but I have still several code paths missing

**Comment by GuiForget** [comment]
@junchuanwang this is ready to be reviewed. One functionality I wanted to add but I'm not sure where to add it is the ability to construct a PathSpecSet from a MaskTree. This functionality would be handy given that the annotation @ProjectionParam is usable only with MaskTree at the moment. Originally I was hoping to add a method `PathSpeSet.of(MaskTree)` but MaskTree class is not available to the `data` module. Do you have a recommended place I could add it?

File: data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java
Comment (BrianPin) [inline data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java:68]: braces on new line?

File: data-transform/src/test/java/com/linkedin/data/transform/filter/TestMaskCreation.java
Comment (BrianPin) [inline data-transform/src/test/java/com/linkedin/data/transform/filter/TestMaskCreation.java:377]: braces on new line?

File: data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java
Comment (GuiForget) [reply data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java:68]: Is there a code style I can refer to?

Comment (BrianPin) [reply data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java:68]: Good q. I think it has been left braces on next line ever since restli first version.
Although I don't think there is strict rules regarding coding style, (for example, the line width)
I just don't want to see in the future there will be many places where mixed type is happening inside the repo

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (BrianPin) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:1]: Might as well add copy-right file headers?

**Review by BrianPin — COMMENTED** [review]
Overall looks good to me, but please change the style to match restli style.

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:38]: remove the _ prefix (not the convention followed in this project)

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:227]: initialize size to parent path size + child path size

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:267]: This should check for the prefix or any parent of the perfix, right?

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:1]: +1. Add copyright comment at the top of all new source files. See existing files for reference.

File: data-transform/src/test/java/com/linkedin/data/transform/filter/TestMaskCreation.java
Comment (evanw555) [reply data-transform/src/test/java/com/linkedin/data/transform/filter/TestMaskCreation.java:377]: +1. This is the convention in Rest.li

File: data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java
Comment (evanw555) [reply data-transform/src/main/java/com/linkedin/data/transform/filter/request/MaskCreator.java:68]: Yeah, AFAIK the Rest.li code style isn't documented formally. It's just based on convention

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:267]: yes you are correct. I'll modify accordingly

**Comment by GuiForget** [comment]
I'll look at the build failure... But all the tests pass locally for me after the rebase

**Comment by evanw555** [comment]
@GuiForget Looks like you've been running into some unfortunate flakiness from the d2/darkcluster tests. You can trigger a re-run in the CI by force-pushing.

**Comment by GuiForget** [comment]
@evanw555 today's rebase passed!

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:34]: For all javadocs in this class, I'd suggest providing examples using the serialized pathspec syntax (e.g. `/foo/bar`, `/foo`) for clarity, especially in more complicated examples (e.g. copy with scope, remove prefix)

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:2]: Should say 2020

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:43]: typo: explicitly

Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:43]: nit: capitalize as "Rest.li"

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: I'm skeptical about using the terminology "full projection", as PathSpecs are not inherently tied to Rest.li projections (that's just one application). However, one could make the argument that the term "projection" is a generic term and isn't implying that it's used for _Rest.li_ projections. What do you think?

I'm also curious to know how a "full projection" PathSpecSet is used. Do you have an example of one use case?

Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: maybe we can change to name "universal". Otherwise the "full" return a empty set and NULL in toArray seems unnatural to me as well

**Comment by junchuanwang** [comment]
I have no other concerns other than the "fullProjection" boolean“ field.

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: In Rest.li world, when you have a PathSpecSet you may have logic that is skipped if the projection doesn't include a specific field:
```
if (pathSpecSet.contains(MyRecord.fields().subrecord()) {
    // do a bunch of stuff that is costly but unnecessary if the field subrecord isn't requested
}
```

And then given the ResourceContext, you can easily build a PathSpecSet with `PathSpecSet.of(resourceContext.getProjectionMask())`. 

When no projection was specified, the ProjectionMask is null and you want to make sure that `contains(PathSpec)` returns true for any PathSpec and that's the meaning of a fullProjection.

Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: For the `toArray()` returning `null`, this is consistent with the behavior of the method `fields(PathSpec[])` in the Rest.li builders.

Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: I'm used to full projection so it is difficult for me to come up with an alternative. I'm not sure universal as an adjective works but I could see `universe` or `universeSet` as suitable. Here are some alternatives that I found in a thesaurus dictionary:
- globalSet
- allInclusiveSet
- blanketSet

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:34]: I edited the javadoc, let me know if this is what you had in mind

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java]: I can't find the typo

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: > When no projection is specified, the ProjectionMask is null and you want to make sure that `contains(PathSpec)` returns true for any PathSpec and that's the meaning of a fullProjection.

In this specific use case it makes sense, but this object is simply a set of path specs. An empty path spec set articulates to a full projection when it's used in the specific application of projection, but path specs aren't only for projections. What does an empty path spec set mean when it's being used for indexing values from a complex data object, or matching ReadOnly/CreateOnly predicates? I don't necessarily think that an empty path spec set contains everything for all use cases.

Comment (junchuanwang) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: thanks for the explanation, I will vote for "allInclusiveSet".

Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:58]: Synced up with @GuiForget about this offline. He clarified that an empty set is not necessarily the same as an "all inclusive" or "full projection" set, which must be created explicitly as such.

As for naming, I say we avoid using the terminology "full projection", and instead use something that implies inclusivity. `allInclusive`/`isAllInclusive` works for me. `complete` may also be good, but I'm not sure if that has a different technical definition when it comes to sets.

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (evanw555) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java:34]: Adding `/foo` directly into code segments may be confusing, as it's not proper Java syntax. However, I still think it's more clear to users. I originally meant adding this in addition to the code examples, like `set of {/a} contains /a` or something like that.

File: data/src/main/java/com/linkedin/data/schema/PathSpecSet.java
Comment (GuiForget) [reply data/src/main/java/com/linkedin/data/schema/PathSpecSet.java]: I renamed fullProjection to allInclusive and propagated the changes throughout the classes and unit tests

**Review by evanw555 — APPROVED** [review]
Thanks for making the suggested changes and thanks for contributing, LGTM. Should probably wait for approval from the other reviewers before merging.

(also nit: but please add an entry to the change for this if you have time)

**Review by BrianPin — APPROVED** [review]

**Pull request merged**
